pkg/pool-weighted/contracts/managed/ManagedPool.sol

```diff
@@ -163,13 +177,11 @@ contract ManagedPool is ManagedPoolSettings {
         uint256 invariantGrowthRatio = WeightedMath._calculateTwoTokenInvariant(
             tokenInWeight,
             tokenOutWeight,
-            FixedPoint.ONE.add(_addSwapFeeAmount(swapRequest.amount).divDown(currentBalanceTokenIn)),
+            FixedPoint.ONE.add(swapRequest.amount.divUp(swapFeeComplement).divDown(currentBalanceTokenIn)),
             FixedPoint.ONE.sub(amountOut.divDown(currentBalanceTokenOut))
         );
 
         _payProtocolAndManagementFees(invariantGrowthRatio);
```

Comment:
I find this to be a bit too hidden - I'm particularly not a fan of the fact that we apply the fee outside and deapply it here.

I'd consider refactoring this so that we return `(amount in, amount out)`, and have this function handle all fees. For the purposes of calling weighted math, it removes the fees, but the full amount in does include it. Then we'd pay protocol fees on the callsite (for both given in and given out). This will also be useful once we add circuit breakers, which will similarly require access to the full amount incl. fees.

Reply:
That'd mean upscaling before applying the fees, but I think this is totally fine. Upscaling is a well-understood operation, and it'll be clearer to always upscale all balances at the same time, instead of the weird thing we have now were we first take fees and then upscale.

It'll also make this more correct: we're currently taking unscaled fees, but deapplying them scaled for external fee collection purposes, leading to some rounding error.